_posts/2016-06-27-bits-of-shapeless-2.md

```diff
@@ -0,0 +1,519 @@
+
+At the end of this post you will have built a basic Json rendering library for sealed hierarchies of case classes. Actually, if you aim to be a rockstar Scala coder you have to do a Json library one way or another :).
+
+It is assumed that the reader knows about json, shapeless heterogeneous lists and how to work with them (see my [last post][firstPost]), knows about implicit resolution and is acquainted with the concept of type classes in Scala.
+
+### Setup
+
+Further from [Shapeless][shapeless] itself, we will use the [JAWN][jawn] Json parsing and `AST` library.
+Some imports are assumed:
+
+```scala
+import shapeless._
+import jawn.ast._
+```
+
+### HLists and beyond
+
+Let's look at some very nice syntax shapeless provides us to manipulate objects as if they were HLists:
+
+```scala
+import syntax.std.tuple._
+// import syntax.std.tuple._
+
+(1,2,3).head
+// res0: Int = 1
+
+(1,2,3).tail
+// res1: (Int, Int) = (2,3)
+```
+
+How is this possible?
+
+```scala
+(1,"a", 1.0).productElements // give me an HList
+// res2: shapeless.::[Int,shapeless.::[String,shapeless.::[Double,shapeless.HNil]]] = 1 :: a :: 1.0 :: HNil
+```
+
+Maybe there is an implicit conversion from tuple to HList involved somehow... But:
+
+```scala
+case class C(x: Int, y: String)
+// defined class C
+
+val c = C(1, "hello")
+// c: C = C(1,hello)
+
+c.length
+// res3: shapeless.Succ[shapeless.Succ[shapeless._0]] = Succ()
+
+c.head
+// res4: Int = 1
+
+c.productElements
+// res5: shapeless.::[Int,shapeless.::[String,shapeless.HNil]] = 1 :: hello :: HNil
+```
+
+As you might know, both tuples and case classes are products, so we may assume that the machinery involved is the same.
+
+Let us go back a bit and explain how all these things were magically converted to HLists:
+
+
+
+### Generic and products
+
+Here I present you the `Generic`type class:
+
+```scala
+trait Generic[T] extends Serializable {
+  type Repr
+  def to(t : T) : Repr
+  def from(r : Repr) : T
+}
+```
+
+What this essentially means is that instances of `Generic[T]` can convert back and fort  from `T`to `Repr` -- a generic representation of `T`. However two questions arise:
+
+1. How do we get instances `of `Generic[T]`, further from implementing them by hand?
+2. And shouldn't be `Repr` some kind of `HList`?
+
+To answer those:
+
+```scala
+val gen = Generic[(Int, String)]
+// gen: shapeless.Generic[(Int, String)]{type Repr = shapeless.::[Int,shapeless.::[String,shapeless.HNil]]} = anon$macro$24$1@70f99ded
+
+gen.to((1, "hello"))
+// res6: gen.Repr = 1 :: hello :: HNil
+```
+
+Note the type of the `gen`variable: the inner type `Repr`is automatically derived for us. This `Generic.Aux` type might be strange for some. If you are not familiar with the Aux pattern, [this is a good explanation][auxPattern], and [this answer explains its applicability  in shapeless][genericAux].
```

Comment:
In the above scala snippet there is no reference to Generic.Aux but to Generic.Repr which is an usage of the Aux pattern